python/taichi/lang/transformer.py

```diff
@@ -423,21 +344,99 @@ def is_decorated(iter):
+            t_loop.iter.args[0] = self.parse_expr(
+                '__ndrange.acc_dimensions[0]')
+            targets = get_targets()
+            targets_tmp = ['__' + name for name in targets]
+            loop_body = t_loop.body
+            for i in range(len(targets)):
+                if i + 1 < len(targets):
+                    stmt = '{} = __I // __ndrange.acc_dimensions[{}]'.format(
+                        targets_tmp[i], i + 1)
+                else:
+                    stmt = '{} = __I'.format(targets_tmp[i])
+                loop_body.append(self.parse_stmt(stmt))
+                stmt = '{} = {} + __ndrange.bounds[{}][0]'.format(
+                    targets[i], targets_tmp[i], i)
+                loop_body.append(self.parse_stmt(stmt))
+                if i + 1 < len(targets):
+                    stmt = '__I = __I - {} * __ndrange.acc_dimensions[{}]'.format(
+                        targets_tmp[i], i + 1)
+                    loop_body.append(self.parse_stmt(stmt))
+            loop_body += node.body
+
+            node = ast.copy_location(t, node)
+            return self.visit(node)  # further translate as a range for
+
+
+        def visit_grouped_ndrange_for(node):
+            from astpretty import pprint
+            # for I in ti.grouped(ti.ndrange(n, m))
+            target = node.target.id
+            template = '''
+if ti.static(1):
+    __ndrange = 0
+    {} = ti.Vector([0] * len(__ndrange.dimensions))
+    for __ndrange_I in range(__ndrange.acc_dimensions[0]):
+        __I = __ndrange_I
+        for __grouped_I in range(len(__ndrange.dimensions)):
+            __grouped_I_tmp = 0
+            if __grouped_I + 1 < len(__ndrange.dimensions):
+                __grouped_I_tmp = 233
             else:
-                elts = node.target.elts
+                __grouped_I_tmp = __I
+            __tmp = 233
+            if __grouped_I + 1 < len(__ndrange.dimensions):
+                __I = 233
+            '''.format(target, target)
+            t = ast.parse(template).body[0]
+            print('ttttttttttttttttttttttttttttttttttttttttttttt')
+            pprint(t)
+            t.body[0].value = node.iter.args[0]
+            t_loop = t.body[2]
+            loop_body = t_loop.body
+            inner_loop_body = loop_body[1].body
+            inner_loop_body[1].body[0].value = self.parse_expr(
+                '__I // __ndrange.acc_dimensions[__grouped_I + 1]')
+            inner_loop_body[2].targets[0] = self.parse_expr(
+                '{}[__grouped_I]'.format(target))
+            inner_loop_body[2].value = self.parse_expr(
+                '__grouped_I_tmp + __ndrange.bounds[__grouped_I][0]')
+            inner_loop_body[3].body[0].value = self.parse_expr(
+                '__I - __grouped_I_tmp * __ndrange.acc_dimensions[__grouped_I + 1]'
+            )
+            loop_body += node.body
+
+            node = ast.copy_location(t, node)
+            return self.visit(node)  # further translate as a range for
```

Comment:
Here `self.visit` will further translate `__grouped_I_tmp = 0` into `__grouped_I_tmp = ti.expr_init(0)` and you get an expr. The solution is to first visit the node body, and then do your own transform so that what you write is what you end up with.

Reply:
After visiting the node body (translating as a range for), how can I locate the statements I want to modify?

Reply:
An easier way: see the trick here and protect your statements with `if ti.static(1)` so that it won't get further translated

https://github.com/taichi-dev/taichi/blob/063542ffe5f16805e9912def6b4e0c0e469faab6/python/taichi/lang/transformer.py#L340

Reply:
Sorry, to be clear: try `__grouped_I_tmp = ti.static(0)` so that this statement won't get transformed into `__grouped_I_tmp = ti.expr_init(0)`.

Reply:
> After visiting the node body (translating as a range for), how can I locate the statements I want to modify?

In my understanding, the only transform you have to do here is to **prepend** some statements to the loop body so that you can get something like `I`. You don't need to modify any existing statements in the old loop body.